src/main/java/com/google/sps/servlets/TripServlet.java

```diff
@@ -18,16 +33,155 @@
 public class TripServlet extends HttpServlet {
 
+  // Create the GeoApiContext object.
+  GeoApiContext context;
+
+  // Constants to get form inputs.
+  private static final String INPUT_TRIP_NAME = "inputTripName";
+  private static final String INPUT_DESTINATION = "inputDestination";
+  private static final String INPUT_DAY_OF_TRAVEL = "inputDayOfTravel";
+
+  // Trip attributes needed to store the Trip Entity in datastore.
+  private String tripName;
+  private String tripDestination;
+  private String tripDayOfTravel;
+  private String destinationName;
+  private String photoSrc;
+
+  @Override
+  public void init() {
+    this.context = new GeoApiContext.Builder()
+      .apiKey(Config.API_KEY)
+      .build();
+  }
+
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     response.setContentType("application/json;");
 
-    // Print out params to site to verify retrieval of "start trip" user input.
-    Enumeration<String> params = request.getParameterNames(); 
-    while (params.hasMoreElements()) {
-      String paramName = params.nextElement();
-      response.getWriter().println(paramName + ": " + request.getParameter(paramName));
+    // Retrieve form inputs to define the Trip object.
+    this.tripName = request.getParameter(INPUT_TRIP_NAME);
+    this.tripDestination = request.getParameter(INPUT_DESTINATION);
+    this.tripDayOfTravel = request.getParameter(INPUT_DAY_OF_TRAVEL);
+
+    // Populate the destinationName and photoSrc fields using Google Maps API.
+    populateDestinationAndPhoto(context, tripDestination);
+
+    // Store the Trip Entity in datastore with the User Entity as an ancestor.
+    storeTripEntity(response, this.tripName, this.destinationName, 
+      this.tripDayOfTravel, this.photoSrc);
+
+    /**
+     * TODO: Remaining code for storing Event and TripDay objects should 
+     * go here, below the above code, as the Trip has to be set first in order
+     * to maintain Entity hierarchy / ancestor paths. 
+     * 
+     * Below methods can also use the field variables fetched from request in 
+     * the above code.
+     */
+
+    // Redirect to the "/trips/" page to show the trip that was added.
+    response.sendRedirect("/trips/");
+  }
+
+  /**
+   * Get the place ID of the text search. Return null if no place ID matches
+   * the search.
+   */ 
+  public String getPlaceIdFromTextSearch(GeoApiContext context, String textSearch) 
+    throws IOException {
+
+    FindPlaceFromTextRequest findPlaceRequest = PlacesApi.findPlaceFromText(context, 
+      textSearch, FindPlaceFromTextRequest.InputType.TEXT_QUERY);
+
+    try {
+      FindPlaceFromText findPlaceResult = findPlaceRequest.await();
+
+      // Return place ID of the first candidate result.
+      if (findPlaceResult.candidates != null) {
+        return findPlaceResult.candidates[0].placeId;
+      }
+      
+      // No candidate is given, so return null.
+      return null;
```

Comment:
In what case would no candidate be given and no exception be thrown? Maybe we should throw an exception if no candidate is given so we don't run into a NullPointerException later on

Reply:
I have a check for if the return is `null` -- check lines 135-138 on `TripServlet.java` (the same file). I'm thinking this could happen if, for some reason, the text search given by the user does not match any places that the backend can find.

While this is admittedly unlikely given I use the same API is an autocomplete, I figured it would probably be good to add a check; for instance, a malicious user could submit a text search that doesn't fit any actual location. I'm open to changing this into throwing an exception as you said though, because in that case we don't have a valid origin / destination.

Should I change this to throw some Exception if no candidate is given?

Reply:
Ok I understand! Returning null makes sense for the way you're using it because you can check if it's null and use a placeholder if so rather than stopping and throwing an exception, but my concern is if this method is used somewhere else (for example getting the place ID for routing), then it would cause a NullPointerException later on. Maybe we could have a default value instead?

Reply:
What sort of default value are you thinking? Also I think `null` might be better, as it lets each method that calls this method `getPlaceIdFromTextSearch(...)` to handle the result a unique way.

Reply:
That reasoning makes sense, I'm fine with leaving this as is then!